Add ServiceProvider base class

markstory opened the pull request:
To provide consistent method name conventions for users we will need a ServiceProvider subclass. This should also help reduce the number of times that users need to import directly from league.

Refs cakephp/docs#6847

markstory committed:
Add ServiceProvider base class

To provide consistent method name conventions for users we will need
a ServiceProvider subclass. This should also help reduce the number of
times that users need to import directly from league.

ADmad commented on src/Core/ServiceProvider.php:
```diff
@@ -0,0 +1,87 @@
+     * Delegate to the bootstrap() method
+     *
+     * This method primarily exists as a shim between the interface
+     * that league/container has and the one we want to offer in CakePHP.
```
Based on this should we mark this method `final` then?

markstory replied:
Nah, if people want to footgun themselves they can.

ADmad replied:
😝

othercorey commented on src/Core/ServiceProvider.php:
```diff
@@ -0,0 +1,87 @@
+    /**
+     * Delegate to the bootstrap() method
+     *
+     * This method primarily exists as a shim between the interface
+     * that league/container has and the one we want to offer in CakePHP.
```
```suggestion
     * This method wraps the league/container function so users
     * only need to use the CakePHP bootstrap() interface.
```

othercorey commented:
Looks ok except the code checks. Can we guarantee the container type is always valid? Then we can suppress that error.

markstory committed:
Fix phpcs

markstory committed:
Ensure interface type

Add runtime checks for interface types and add psalm supression as
I don't know how to appease it properly.

robertpustulka commented on tests/test_app/TestApp/ServiceProvider/PersonServiceProvider.php:
```diff
@@ -0,0 +1,22 @@
+
+class PersonServiceProvider extends ServiceProvider
+{
+    protected $provides = ['boot', 'sally'];
```
I don't like this being a property in League DIC, not very intuitive. Can we have a method for this?

ADmad replied:
Based on the docs https://container.thephpleague.com/3.x/service-providers/  it doesn't seem there's any alternative to populating this array. So we would have to add a method ourselves if needed.

robertpustulka replied:
That's what I had in mind.

markstory replied:
Would the method return the list of services provides? That will end up running a bit slower than a property would as the array would be re-allocated many times during a request.

markstory committed:
Fix php7.2 incompatibility

markstory committed:
Fix phpcs

markstory committed:
Apply suggestions from code review

Co-authored-by: othercorey <corey.taylor.fl@gmail.com>
